File: packages/nim/oauth/src/oauth_do.nim
## OAuth device flow SDK for .do APIs.

import std/[httpclient, json, os, times]
import oauth_do/device
import oauth_do/storage

export device, storage

const
  DefaultClientId* = "client_01JQYTRXK9ZPD8JPJTKDCRB656"

type
  OAuthClient* = object
    clientId: string
    deviceFlow: DeviceFlow
    storage: TokenStorage

proc newOAuthClient*(clientId: string = DefaultClientId): OAuthClient =
  ## Creates a new OAuth client.
  OAuthClient(
    clientId: clientId,
    deviceFlow: newDeviceFlow(clientId),
    storage: newTokenStorage()
  )

proc authorizeDevice*(client: OAuthClient, scope: string = "openid profile email"): DeviceAuthResponse =
  ## Initiates device authorization flow.
  client.deviceFlow.authorize(scope)

proc pollForTokens*(client: OAuthClient, deviceCode: string, interval: int, expiresIn: int): TokenResponse =
  ## Polls for tokens after user authorizes the device.
  let tokens = client.deviceFlow.pollForTokens(deviceCode, interval, expiresIn)
  client.storage.saveTokens(tokens)
  tokens

proc getUser*(client: OAuthClient, accessToken: string = ""): JsonNode =
  ## Gets current user info using stored or provided access token.
  var token = accessToken
  if token == "":
    token = client.storage.getAccessToken()
  if token == "":
    raise newException(ValueError, "No access token available")
  client.deviceFlow.getUserInfo(token)

proc login*(client: OAuthClient, scope: string = "openid profile email"): TokenResponse =
  ## Interactive login flow.
  let deviceInfo = client.authorizeDevice(scope)

  echo "\nTo sign in, visit: ", deviceInfo.verificationUri
  echo "And enter code: ", deviceInfo.userCode, "\n"

  let interval = if deviceInfo.interval > 0: deviceInfo.interval else: 5
  let expiresIn = if deviceInfo.expiresIn > 0: deviceInfo.expiresIn else: 900

  client.pollForTokens(deviceInfo.deviceCode, interval, expiresIn)

proc logout*(client: OAuthClient) =
  ## Logs out by removing stored tokens.
  client.storage.deleteTokens()

proc hasTokens*(client: OAuthClient): bool =
  ## Checks if tokens exist.
  client.storage.hasTokens()

File: packages/elixir/oauth/lib/oauth_do.ex
defmodule OauthDo do
  @moduledoc """
  OAuth device flow SDK for .do APIs.
  """

  alias OauthDo.{Device, Storage}

  @doc """
  Initiates device authorization flow.
  Returns device code info including user_code and verification_uri.
  """
  def authorize_device(opts \\ []) do
    Device.authorize(opts)
  end

  @doc """
  Polls for tokens after user authorizes the device.
  """
  def poll_for_tokens(device_code, interval, expires_in, opts \\ []) do
    Device.poll_for_tokens(device_code, interval, expires_in, opts)
  end

  @doc """
  Gets the current user info using the stored access token.
  """
  def get_user(access_token \\ nil) do
    token = access_token || Storage.get_access_token()

    case token do
      nil ->
        {:error, :no_token}

      token ->
        case Req.get("https://apis.do/me",
               headers: [{"authorization", "Bearer #{token}"}]
             ) do
          {:ok, %{status: 200, body: body}} -> {:ok, body}
          {:ok, %{status: status}} -> {:error, {:http_error, status}}
          {:error, reason} -> {:error, reason}
        end
    end
  end

  @doc """
  Interactive login flow. Prints instructions and waits for authorization.
  """
  def login(opts \\ []) do
    case authorize_device(opts) do
      {:ok, device_info} ->
        IO.puts("\nTo sign in, visit: #{device_info["verification_uri"]}")
        IO.puts("And enter code: #{device_info["user_code"]}\n")

        poll_for_tokens(
          device_info["device_code"],
          device_info["interval"] || 5,
          device_info["expires_in"] || 900,
          opts
        )

      {:error, reason} ->
        {:error, reason}
    end
  end

  @doc """
  Logs out by removing stored tokens.
  """
  def logout do
    Storage.delete_tokens()
  end
end
